File: arcgis_map_android/android/src/main/kotlin/esri/arcgis/flutter_plugin/ArcgisMapService.kt
package esri.arcgis.flutter_plugin

import com.esri.arcgisruntime.concurrent.Job
import com.esri.arcgisruntime.tasks.vectortilecache.ExportVectorTilesJob
import com.esri.arcgisruntime.tasks.vectortilecache.ExportVectorTilesTask
import esri.arcgis.flutter_plugin.model.ExportVectorTilesParametersPayload
import esri.arcgis.flutter_plugin.model.toExportVectorTilesParameters
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class ArcgisMapService(binaryMessenger: BinaryMessenger) :
    MethodChannel.MethodCallHandler {
    // Key is the id of the task, the value is the actual task
    private var exportVectorTilesTasks = mutableMapOf<String, ExportVectorTilesTask>()

    // Key is the id of the task that created the job, the value is the actual job.
    private var exportVectorTilesJobs = mutableMapOf<String, ExportVectorTilesJob>()

    private var methodChannel: MethodChannel =
        MethodChannel(binaryMessenger, "esri.arcgis.flutter_plugin")

    init {
        methodChannel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "create_export_vector_tiles_task" -> onCreateExportVectorTilesTask(call, result)
            "start_export_vector_tiles_task_job" -> onStartExportVectorTilesTaskJob(call, result)
            "cancel_export_vector_tiles_task_job" -> onDisposeJob(call, result)
            "cancel_export_vector_tiles_task_jobs" -> onDisposeJobs(result)
        }
    }

    private fun onDisposeJobs(result: MethodChannel.Result) {
        disposeExportVectorTileJobs()
        result.success(null)
    }

    private fun onDisposeJob(call: MethodCall, result: MethodChannel.Result) {
        val map = call.arguments as Map<String, Any>
        val taskId = map["taskId"] as String

        val job = exportVectorTilesJobs[taskId]
        if (job == null) {
            result.error("job_not_found", "Job with id $taskId was not found.", null)
            return
        }

        val success = job.cancel()
        if (!success) {
            result.error("job_cancel_failed", "Canceling job failed.", null)
            return
        }

        result.success(null)
    }


    private fun onCreateExportVectorTilesTask(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val args = call.arguments as Map<String, Any>
        val url = args["url"] as String
        val taskId = args["id"] as String
        val exportTask = ExportVectorTilesTask(url)
        exportVectorTilesTasks[taskId] = exportTask
        result.success(null)
    }

    private fun onStartExportVectorTilesTaskJob(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val args = call.arguments as Map<String, Any>
        val taskId = args["taskId"] as String
        val exportParametersMap = args["exportVectorTilesParameters"] as Map<String, Any>
        val parameters = exportParametersMap.parseToClass<ExportVectorTilesParametersPayload>()
            .toExportVectorTilesParameters()
        val vectorTileCachePath = args["vectorTileCachePath"] as String
        val task = exportVectorTilesTasks[taskId]!!
        val job = task.exportVectorTiles(parameters, vectorTileCachePath)

        // Cancel the existing job before we overwrite it.
        exportVectorTilesJobs[taskId]?.cancel()
        exportVectorTilesJobs[taskId] = job

        val progressListener = {
            methodChannel.invokeMethod(
                "export_vector_tiles_job_progress",
                mapOf("taskId" to taskId, "progress" to job.progress.toDouble())
            )
        }

        lateinit var jobChangedListener: Runnable
        jobChangedListener = Runnable {
            when (job.status) {
                Job.Status.FAILED -> {
                    job.removeProgressChangedListener(progressListener)
                    job.removeJobChangedListener(jobChangedListener)
                    result.error(
                        "download_failed",
                        job.error.message + " error code: ${job.error.errorCode}",
                        job.error.additionalMessage
                    )
                }

                Job.Status.SUCCEEDED -> {
                    job.removeProgressChangedListener(progressListener)
                    job.removeJobChangedListener(jobChangedListener)
                    result.success(null)
                }

                else -> {}
            }
        }

        job.addProgressChangedListener(progressListener)
        job.addJobChangedListener(jobChangedListener)
        job.start()
    }

    fun disposeExportVectorTileJobs() {
        val idsToRemove = mutableListOf<String>()
        exportVectorTilesJobs.entries.forEach { (taskId, job) ->
            job.cancel()
            idsToRemove.add(taskId)
        }
        idsToRemove.forEach(exportVectorTilesJobs::remove)
    }
}

File: arcgis_map_ios/ios/Classes/ArcgisMapService.swift
//
//  ExportVectorTilesTaskFactory.swift
//  arcgis_map_ios
//
//  Created by Julian Bissekkou on 10.01.23.
//

import Foundation
import ArcGIS

class ArcgisMapService {
    
    // Key is the id of the task, the value is the actual task
    private var exportVectorTilesTasks = [String: AGSExportVectorTilesTask]()
    
    // Key is the id of the task that created the job, the value is the actual job.
    private var exportVectorTilesJobs = [String: AGSExportVectorTilesJob]()
    
    private let methodChannel: FlutterMethodChannel
    
    init(methodChannel: FlutterMethodChannel) {
        self.methodChannel = methodChannel
    }
    
    public func onCreateExportVectorTilesTask(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let dict = call.arguments! as! Dictionary<String, Any>
        let url = dict["url"] as! String
        let taskId = dict["id"] as! String
        
        let exportTask = AGSExportVectorTilesTask(url: URL(string: url)!)
        exportVectorTilesTasks[taskId] = exportTask
        result(nil)
    }
    
    public func onStartExportVectorTileTaskJob(_ call: FlutterMethodCall, _ result:@escaping FlutterResult) {
        let dict = call.arguments! as! Dictionary<String, Any>
        let taskId = dict["taskId"] as! String
        let exportParamsDict = dict["exportVectorTilesParameters"] as! Dictionary<String, Any>
        let exportParams: ExportVectorTilesParametersPayload = try! JsonUtil.objectOfJson(exportParamsDict)
        let vectorTileCachePath = dict["vectorTileCachePath"] as! String
        let task = exportVectorTilesTasks[taskId]!
        
        let job = task.exportVectorTilesJob(
            with: exportParams.toAgsExportVectorTileParameters(),
            downloadFileURL: URL(string: vectorTileCachePath)!
        )
        // Cancel the existing job before we overwrite it.
        if let existingJob = exportVectorTilesJobs[taskId] {
            existingJob.cancel {_ in }
        }
        exportVectorTilesJobs[taskId] = job
        
        job.start(
            statusHandler: { status in
                let progress = Double(job.progress.completedUnitCount) / Double(job.progress.totalUnitCount)
                
                self.methodChannel.invokeMethod(
                    "export_vector_tiles_job_progress",
                    arguments: ["taskId" : taskId, "progress" : progress]
                )
                
                // TODO how to notify when job is cancled?
                switch(status) {
                case .succeeded:
                    result(nil)
                case .failed:
                    result(FlutterError(code: "download_failed", message: "Download failed \(job.error?.localizedDescription)", details: nil))
                @unknown default:
                    print("\(status)")
                }
            },
            completion: { data, error in
                if let error = error {
                    result(FlutterError(code: "export_task_error", message: error.localizedDescription, details: nil))
                } else {
                    result(nil)
                }
            }
        )
    }
    
    public func cancelJob(_ call: FlutterMethodCall, _ result:@escaping FlutterResult) {
        let dict = call.arguments as! Dictionary<String, Any>
        let taskId = dict["taskId"] as! String
        
        guard let job = exportVectorTilesJobs[taskId] else {
            result(FlutterError(code: "task_not_found", message: "Task with id\(taskId) not found", details: nil))
            return
        }
        
        
        job.cancel { error in
            guard let error = error else {
                result(nil)
                return
            }
            result(FlutterError(code: "job_cancel_failed", message: "Canceling job failed: \(error.localizedDescription)", details: nil))
        }
    }
    
    public func cancelJobs(_ result:@escaping FlutterResult) {
        exportVectorTilesJobs.forEach { taskId, job in
            job.cancel { _ in
                self.exportVectorTilesJobs[taskId] = nil
            }
        }
        
        result(nil)
    }
}


struct ExportVectorTilesParametersPayload: Codable {
    let areaOfInterest: EnvelopePayload
    let maxLevel: Int
}

extension ExportVectorTilesParametersPayload {
    func toAgsExportVectorTileParameters() -> AGSExportVectorTilesParameters {
        let params = AGSExportVectorTilesParameters()
        params.maxLevel = self.maxLevel
        params.areaOfInterest = self.areaOfInterest.toAGSEnvelope(spatialReference: .webMercator())
        
        return params
    }
}
